go/upgrade/api/api.go

```diff
@@ -62,28 +62,120 @@ var (
 	ErrUpgradeInProgress = errors.New(ModuleName, 6, "upgrade: can not cancel upgrade in progress")
 )
 
+// OwnHash returns the hash of the executable that started the current process.
+func OwnHash() (*hash.Hash, error) {
+	path, err := os.Executable()
+	if err != nil {
+		return nil, err
+	}
+
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+
+	h := hash.NewFromBytes(contents)
+	return &h, nil
+}
+
+// UpgradeMethod is an upgrade descriptor method.
+type UpgradeMethod uint8
+
+const (
+	// UpgradeMethodInternal is the internal upgrade method, where the node
+	// binary itself has the migration code.
+	UpgradeMethodInternal UpgradeMethod = 1
+
+	// UpgradeMethodInternalName is the name of the upgrade method.
+	UpgradeMethodInternalName = "internal"
+)
+
+// String returns a string representation of a UpgradeMethod.
+func (m UpgradeMethod) String() string {
+	switch m {
+	case UpgradeMethodInternal:
+		return UpgradeMethodInternalName
+	default:
+		return fmt.Sprintf("[unknown upgrade method: %d]", m)
+	}
+}
+
+// MarshalText encodes a UpgradeMethod into text form.
+func (m UpgradeMethod) MarshalText() ([]byte, error) {
+	switch m {
+	case UpgradeMethodInternal:
+		return []byte(UpgradeMethodInternalName), nil
+	default:
+		return nil, fmt.Errorf("invalid upgrade method: %d", m)
+	}
+}
+
+// UnmarshalText decodes a text slice into an UpgradeMethod.
+func (m *UpgradeMethod) UnmarshalText(text []byte) error {
+	switch string(text) {
+	case UpgradeMethodInternalName:
+		*m = UpgradeMethodInternal
+	default:
+		return fmt.Errorf("invalid upgrade method: %s", string(text))
+	}
+	return nil
+}
+
 // Descriptor describes an upgrade.
 type Descriptor struct {
```

Comment:
What would be the use case for it? In case we'd like to use it as an option instead of the current `Identifier` field (so we wouldn't need exact hash matches) that would require some more refactoring of the `upgrade` module (which currently assumes that `UpgradeInternal` method uses the hashes). Probably a new `UpgradeMethod` type?

Reply:
Yeah so I was thinking that the hash of a binary is not the best thing, for example in a world where you have different implementations. Of course it could be "hash of an externally-defined manifest" as well and that manifest would have the details. So maybe we can leave it as-is and use a different method which defines the identifier as "hash of a manifest".